### DISCUSSION TITLE
Docker YOURLS_USER Multiple Entries

### DISCUSSION THREAD

**Discussion by PES-Dev1** [YOURLS/YOURLS#3428] [Q&A]
I am using the default YOURLS image to create a Service in a Docker cluster. Currently, the service is scaled to three individual nodes attached to the same MySQL DB. The system works as intended. At present, there is only one account that has admin access to add, edit, or delete entries. I would like to add additional YOURLS_USER entries for additional 'admins' to manage entries. I tried adding an additional YOURLS_USER to the Docker Service config, but the first YOURLS_USER was ignored and only the second entry could log-in as 'Admin'. (It might be more accurate to say the first entry was replaced by the second entry.)

I tried several times to tweak the Docker Service string to match the $yourls_user_passwords in the config.php without any success. (This may have been wishful thinking on my part.)

My question is this, is it possible to dynamically add YOURLS_USER accounts when it is configured as a Docker Service? I thought about trying a volume mount for the config.php file, but I was not sure if that would be practical.

**Comment by LeoColomb** [comment]
> I thought about trying a volume mount for the config.php file, but I was not sure if that would be practical.

@PES-Dev1 Your thought is correct! 😄
Not as practical as environment variables, but there is no alternative for multiple user.

**Comment by logusivam** [comment]
This is a fundamental limitation of how YOURLS handles users, not a Docker configuration problem. Understanding why your second entry replaced the first is the key to understanding the correct solution.

---

## Why only the second `YOURLS_USER` worked

The official YOURLS Docker image maps environment variables to `user/config.php` at container startup. `YOURLS_USER` and `YOURLS_PASS` are singular environment variables — they map to exactly one `$yourls_user_passwords` entry in `config.php`. When you define two `YOURLS_USER` entries in your Docker service config, Docker only keeps the last definition of any duplicate environment variable key. So the second entry doesn't add to the first — it overwrites it entirely. This is a Docker (and Linux environment variable) constraint, not a YOURLS one.

---

## How YOURLS actually stores multiple admin users

YOURLS's multi-user configuration uses a PHP array in `user/config.php`, not a single key-value pair:

```php
$yourls_user_passwords = [
    'alice'   => 'hashed_password_1',
    'bob'     => 'hashed_password_2',
    'charlie' => 'hashed_password_3',
];
```

There is no environment variable syntax that maps to a PHP array with multiple entries — you cannot express this through `YOURLS_USER`/`YOURLS_PASS` env vars no matter how you combine them. The Docker image's entrypoint script simply doesn't support that mapping.

---

## The correct solution — volume-mount a custom `config.php`

Your instinct about the volume mount was exactly right, and it is the practical, standard approach for this use case. Here's how to do it properly.

### Step 1 — Create your custom `config.php`

Start from the YOURLS sample config file (`user/config-sample.php` from the YOURLS repo) and fill in your values. The key section for multiple users:

```php
<?php
// ** Database settings ** //
define( 'YOURLS_DB_HOST',   'your_mysql_host' );
define( 'YOURLS_DB_USER',   'your_db_user' );
define( 'YOURLS_DB_PASS',   'your_db_password' );
define( 'YOURLS_DB_NAME',   'yourls' );
define( 'YOURLS_DB_PREFIX', 'yourls_' );

// ** Site settings ** //
define( 'YOURLS_SITE',      'https://yourdomain.com' );
define( 'YOURLS_LANG',      '' );
define( 'YOURLS_UNIQUE_URLS', true );
define( 'YOURLS_PRIVATE',   true );

// ** Multiple admin users ** //
$yourls_user_passwords = [
    'alice'   => 'plaintext_or_phpass_hash_1',
    'bob'     => 'plaintext_or_phpass_hash_2',
    'charlie' => 'plaintext_or_phpass_hash_3',
];

// ** URL shortening settings ** //
define( 'YOURLS_COOKIEKEY', 'put_a_random_string_here' );
```

> ⚠️ For passwords, you can use plaintext during initial setup — YOURLS will hash them automatically on first use and rewrite the config file. However in a Docker volume-mount setup the rewrite may not persist across container restarts if the mount is read-only. The safer approach is to pre-hash passwords using YOURLS's phpass format. You can generate hashes by temporarily running YOURLS normally, logging in once with each account, then copying the resulting hashed values from the generated `config.php`.

### Step 2 — Place the config file on your Docker host

```bash
mkdir -p /srv/yourls/config
cp config.php /srv/yourls/config/config.php
chmod 640 /srv/yourls/config/config.php
```

### Step 3 — Mount it in your Docker service

**Docker Compose:**

```yaml
services:
  yourls:
    image: yourls:latest
    volumes:
      - /srv/yourls/config/config.php:/var/www/html/user/config.php:ro
    environment:
      # Remove YOURLS_USER and YOURLS_PASS entirely — config.php handles auth now
      YOURLS_DB_HOST: your_mysql_host
      YOURLS_DB_USER: your_db_user
      YOURLS_DB_PASS: your_db_password
      YOURLS_DB_NAME: yourls
      YOURLS_SITE:    https://yourdomain.com
    # ... rest of your service config
```

**Docker Swarm service (your current setup):**

```bash
docker service update \
  --mount-add type=bind,source=/srv/yourls/config/config.php,target=/var/www/html/user/config.php,readonly \
  yourls_service_name
```

> ⚠️ In a three-node Docker Swarm cluster, the mounted file must exist at the **same path on every node** that can run the service. Use one of:
> - A shared network filesystem (NFS, GlusterFS, Ceph) mounted at the same path on all nodes
> - Docker Swarm configs (see below — this is the cleaner Swarm-native approach)
> - A constraint to pin the service to a specific node (defeats the point of three nodes but works for testing)

---

## The cleanest Swarm-native approach — Docker Configs

Docker Swarm has a built-in mechanism for exactly this use case: **Docker Configs**. These store file content in the Swarm's distributed state and mount it into containers automatically on every node, with no shared filesystem required.

```bash
# Create the config from your file
docker config create yourls_config /srv/yourls/config/config.php

# Add it to your service
docker service update \
  --config-add source=yourls_config,target=/var/www/html/user/config.php \
  yourls_service_name
```

To update users later:

```bash
# Create a new version of the config
docker config create yourls_config_v2 /srv/yourls/config/config.php

# Update the service to use the new version
docker service update \
  --config-rm yourls_config \
  --config-add source=yourls_config_v2,target=/var/www/html/user/config.php \
  yourls_service_name
```

> ⚠️ Docker Configs are **immutable** — you can't edit one in place. You must create a new config with a new name (e.g. `yourls_config_v2`) and update the service to use it. The old config can then be removed once no service references it.

---

## Important: remove `YOURLS_USER` and `YOURLS_PASS` env vars

Once you're using a volume-mounted or Swarm Config `config.php`, **remove** `YOURLS_USER` and `YOURLS_PASS` from your Docker service environment variables entirely. The entrypoint script writes those env vars into `config.php` at startup — if it runs and overwrites your mounted file's user section, your multi-user setup will be silently replaced with a single-user config again.

Check your YOURLS image's entrypoint script behavior: some versions only write `config.php` if it doesn't already exist (safe for volume mounts), others always overwrite (problematic). If your image always overwrites, using `:ro` (read-only) on the bind mount will prevent the overwrite — the entrypoint will fail to write to it, but as long as your `config.php` is already complete, YOURLS itself will work fine.

---

## Summary

| Approach | Works for 3-node Swarm? | Complexity |
|----------|------------------------|------------|
| Multiple `YOURLS_USER` env vars | ❌ No — Docker only keeps last duplicate | N/A |
| Volume bind-mount `config.php` | ✅ Yes, with shared filesystem on all nodes | Medium |
| Docker Swarm Config | ✅ Yes — Swarm-native, no shared filesystem needed | Low once set up |

Use **Docker Swarm Configs** — it is the correct, production-grade solution for your three-node cluster and requires no shared filesystem infrastructure.

Hope this helps — if it answers your question, would you mind clicking **"Mark as answer"** so others searching for the same thing can find it quickly?
